kythe/go/util/compare/compare.go

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2015 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package compare implements comparisons between Kythe values.
+package compare
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	spb "kythe.io/kythe/proto/storage_go_proto"
+)
+
+// An Option changes the behavior of the generic comparisons.
+type Option interface{ isOption() }
+
+// And is an Option that unions other Options.
+func And(opt ...Option) Option { return and(opt) }
+
+type and []Option
+
+func (and) isOption() {}
+
+// With is an Option that provides a custom comparison function for the final
+// values being compared.
+type With func(a, b interface{}) Order
+
+func (With) isOption() {}
+
+type reverseOrder struct{}
+
+func (reverseOrder) isOption() {}
+
+// Reversed returns an Option that reverses the resulting Order from a
+// comparison.
+func Reversed() Option { return reverseOrder{} }
+
+// By is an Option that transforms a value before performing a comparison.  It
+// should work on both sides of the comparison equivalently.
+type By func(interface{}) interface{}
+
+func (By) isOption() {}
+
+// Compare returns the Order between two arbitrary values of the same type.
+//
+// Only the following types are currently supported: {string, int, []byte}.
+//
+// Options may be provided to change the semantics of the comparison.  Other
+// types may be compared if an appropriate By Option transforms the values into
+// a supported type or a With Option is provided for the types given.
+//
+// Note: this function panics if a and b are different types
+func Compare(a, b interface{}, opts ...Option) (o Order) {
```

Comment:
It might be nice to note that only the final With clause is honored.

Reply:
I added the note to the With doc.